core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -827,159 +848,170 @@ public T next() {
             }
         }
 
+        public int getNextKeyIndex() {
+            return nextIndex;
+        }
+
         // the actual next()
         abstract T internalNext();
 
         /**
          * Advances next to higher entry.
+         * Return previous index
          */
-        void advance() {
-            if (next == Chunk.NONE)
+        Pair<ByteBuffer, Handle> advance() {
+
+            Object lastKey = lastKeyBeforeChunkDelete.get();
+            if (!(lastKey instanceof Chunk)) {
+                //Current chunk got released
+                if (isDescending) {
+                    hiInclusive = true;
+                    hi = (K)lastKeyBeforeChunkDelete.get();
+                }
+                else {
+                    loInclusive = true;
+                    lo = (K)lastKeyBeforeChunkDelete.get();
+                }
+                initChunk();
+                if (nextChunkIter == null) {
+                    throw new NoSuchElementException();
+                }
+                if (nextIndex == Chunk.NONE) {
+                    throw new ConcurrentModificationException();
```

Comment:
add message in exception something like "next key may have been deleted while iterating the keys"